ESLint Additional Rules (#6045)
department-of-veterans-affairs/va.gov-team: platform/engineering/frontend/eslint/eslint-testing.md

> # ESLint Rule Changes and Testing

Reviewer: What do you think about calling this file something like `adding-new-rules.md`?

**Author:** Jhonny Gonzalez
**Last Updated:** February 21, 2020
**eslint version:** 4.18.2

## Overview

This is an initial process to test and deploy new rules for ESLint.

## Objective

> To improve the code coverage and code quality in the repo by enforcing new ESLint rules. This also will help all VFS teams to write better code.

Reviewer: ```suggestion
To improve code quality in the repo by enforcing new ESLint rules.
```

Unless you think it's actually increasing the _testing_ coverage.

## Background

> We installed ESLint a couple of years ago in the VA repo and hasn't been updated nor improved since then.

Reviewer: ```suggestion
We installed ESLint a couple of years ago in the `vets-website` repo and hasn't been updated nor improved since then.
```

> Now, that VFS teams will have an increased ownership on their pull request reviews, this will help them getting a better judgement as well as facilitate the review process.

Reviewer: ```suggestion
Now that VFS teams will have an increased ownership on their pull request reviews, this will help them get a better judgement as well as facilitate the review process.
```

I'm leaning toward something more like
> The Veteran-facing Service Platform (VSP) has been reviewing all Veteran-facing Services Application (VSA) teams' Pull Requests (PRs) on `vets-website`. Because of the limited nature of the PR reviews VSP engineers can give, one of the main types of feedback we have been able to give concerned code quality.
>
> Because VSP is moving away from reviewing all VSA PRs, we plan to better leverage ESLint to automate that code quality feedback.

What do you think?

Reviewer 2: I love it! Definitely better explained.

# Specifics

> ## ESLintin Procedure

Reviewer: ```suggestion
## Adding New Rules
```
This is more about how to add new rules, right?

Reviewer 2: right

> The ESLinting rules in CircleCI will be used as an experimental staging area as we explore new rules and ways to communicate any issues that are discovered.

Reviewer: ```suggestion
The ESLint rules in CircleCI will be used as an experimental staging area as we explore new rules and ways to communicate any issues that are discovered.
```

Probably worth mentioning here that there will be two tasks in CircleCI. One that runs rules which are enforced, and one that runs rules which are experimental. It's this experimental workflow that we're talking about primarily here.

Reviewer 2: Sounds good!

## Code Location

### ESLint main file

`.eslintrc` => located at the root folder.

### ESLint Testing file

`circle.eslint.json` => rules under review will be located here

## Plugin/New Rules Audit

> Before contemplating the idea of adding a new plugin with rules, it is important to perform an audit. This audit will help us determine if the current repo is in compliance with each rule. All outstanding errors from rules must be cleared (disabled or fixed) before going into the testing stage.

Reviewer: ```suggestion
Before adding a new plugin or new rules, it is important to perform an audit. This audit will help us determine if the current repo is in compliance with each rule. All outstanding errors from rules must be cleared (disabled or fixed) before going into the testing stage.
```
So the important part here are the _rules_ we're running, not just the plugins. We can disable rules on new plugins, so I think it's important to treat each rule individually.

> This audit will help us determine if the current repo is in compliance with each rule.

Are we trying to determine if we're in compliance with each _new_ rule, or each _existing_ rule?

> All outstanding errors from rules must be cleared (disabled or fixed) before going into the testing stage.

Same question: Are these errors from the new rules we want to enforce?

Reviewer 2: Got it. Let me phrase it better.

VFS teams are responsible for fixing the errors generated by the rule under review. Once all the errors have been cleared, we can proceed by adding the rule(s) to the testing stage file (`circle.eslint.json`)

## Rules acceptance

> All rules under review will always be considered to become enforced. However, based on feedback or complexity they can become disabled and not enforced.

Reviewer: It might be clearer to explain this in terms of the rule's lifecycle. So like:
1. Rule X is considered
2. Rule X is added to the unenforced ruleset
3. Rule X is test run in the unenforced ruleset for two weeks
4. Rule X is added to the enforced ruleset
    - Unless there are unresolved errors
    - Unless there has been significant pushback against the rule and VSP decides to not enforce it

Reviewer 2: Wow, I like this

## Feedback

> VFS are encouraged to provide feedback from any rule that is under review. To do so, please use the tag `@frontend-review-group` or message any of the front-end tools team members in Slack and provide us with your feedback.

Reviewer: I like this. :+1:
